### ISSUE TITLE
TypeError: type object argument after ** must be a mapping, not list

### ISSUE THREAD

**Issue by hoyeunglee** [explosion/spaCy#1680]
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "C:\Python27\lib\site-packages\spacy\language.py", line 391, in update
    gold = GoldParse(doc, **gold)
TypeError: type object argument after ** must be a mapping, not list

`train_data = []
for bb in content[1:100]:
    ipPattern = re.compile('\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}')
    findIP = re.search(ipPattern, bb.decode('utf-8'))
    if findIP:
        #train_data = train_data + [(bb.decode('utf-8'), [(findIP.start(), findIP.end(), 'IPADDRESS')])]
        train_data = train_data + [(bb.decode('utf-8'),  [(findIP.start(), findIP.end(), 'IPADDRESS')])]
        print("ok")

nlp = spacy.blank('en')
optimizer = nlp.begin_training()
for i in range(20):
    random.shuffle(train_data)
    for text, annotations in train_data:
        nlp.update([text], [annotations], sgd=optimizer)`

**Comment by akshay-1993** [comment]
I was facing the same issue, but went through the examples on spacy.io for training entity, it helps.
I faced this because i was passing list not map, same with your code.

Since list can be modified, we need to pass maps only.

**Comment by hoyeunglee** [comment]
I follow the format of train data in official web site but got error, and there are hash table and dictionary , which kind of map should be used?

**Comment by hoyeunglee** [comment]
Is there any symbols or format to use when using map?

**Comment by akshay-1993** [comment]
Dictionary we can use

**Comment by hoyeunglee** [comment]
Thanks, I will try in office tomorrow.

**Comment by ines** [comment]
Your code example is a bit hard to read, but I think the problem here is that you need to pass in a dictionary like `{'entities': []}` as the annotations, not just a list of entities. `nlp.update` lets you update all kinds of annotations, so you need to tell spaCy explicitly that your annotations are entity annotations – and not tags or dependency labels.

So your training data should look something like this.

```python
train_data = [
    ("127.0.0.1 is an IP address", {'entities': [(0, 9, 'IPADDRESS')] })
]
```

**Comment by hoyeunglee** [comment]
error after using dictionary

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "C:\Python27\lib\random.py", line 291, in shuffle
    x[i], x[j] = x[j], x[i]
KeyError: 43

```
def addtogroupkey(keyandmemory, key1, memorycontent):
    k = key1
    if k in keyandmemory: 
        keyandmemory[k].append(memorycontent) 
    else: 
        keyandmemory[k] = [memorycontent]
    return keyandmemory

train_data = {}
for bb in content[1:100]:
    ipPattern = re.compile('\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}')
    findIP = re.search(ipPattern, bb.decode('utf-8'))
    if findIP:
        #train_data = train_data + [(bb.decode('utf-8'), [(findIP.start(), findIP.end(), 'IPADDRESS')])]
        #train_data = train_data + [(bb.decode('utf-8'),  [(findIP.start(), findIP.end(), 'IPADDRESS')])]
        train_data = addtogroupkey(train_data, bb.decode('utf-8'), [(findIP.start(), findIP.end(), 'IPADDRESS')])
        print("ok")
```

https://gist.github.com/hoyeunglee/550e60a0610a417611f1fad0e7299312

**Comment by ines** [comment]
It's still a bit hard to follow, since your code is quite specific, but the main problems I see here is:

Your `train_data` is a dictionary – but it needs to be a list of tuples. This is why `random.shuffle` fails to shuffle it, because it expects a list, not a dict. The train data format should look like this:

```python
train_data = [("127.0.0.1 is an IP address", {'entities': [(0, 9, 'IPADDRESS')] })]
```

Or, using your code:

```python
[
    (bb.decode('utf-8'), {'entities': [(findIP.start(), findIP.end(), 'IPADDRESS')]})
    #    ^ the text     ^  dict of entities  ^ start index     ^ end index    ^  label
]
```

**Comment by hoyeunglee** [comment]
thank succeed to train
